dashboard/lib/pd/foorm/survey_reporter.rb

```diff
@@ -54,34 +31,97 @@ def self.get_workshop_report(workshop_id)
+      )
+      # get overall rollup
+      overall_rollup = get_rollup_for_course(ws_data.course, rollup_question_details, facilitators)
+      overall_rollup_per_facilitator = facilitators ?
+                                         get_facilitator_rollup_for_course(
+                                           facilitators,
+                                           ws_data.course,
+                                           rollup_question_details
+                                         ) :
+                                         {}
 
-      result_data[:workshop_rollups][:single_workshop] = {
-        averages: rollup[:averages],
-        response_count: rollup[:response_count],
-        workshop_id: ws_data.id
-      }
+      result_data[:workshop_rollups] = {}
 
-      # get overall rollup
-      overall_rollup = get_rollup_for_course(ws_data.course, rollup_question_details)
-      result_data[:workshop_rollups][:overall] = {
-        averages: overall_rollup[:averages],
-        response_count: overall_rollup[:response_count]
-      }
+      rollup_question_details.each do |key, questions|
+        result_data[:workshop_rollups][key] = {
+          questions: questions,
+          single_workshop: rollup[key],
+          overall_facilitator: facilitators ? overall_rollup_per_facilitator[key] : {},
+          overall: overall_rollup[key]
+        }
+      end
 
       result_data
     end
 
     # Get rollup for all survey results for the given course
-    def self.get_rollup_for_course(course_name, rollup_question_details)
+    def self.get_rollup_for_course(course_name, rollup_question_details, facilitators)
       workshop_ids = Pd::Workshop.where(course: course_name).where.not(started_at: nil, ended_at: nil).pluck(:id)
-      ws_submissions, form_submissions, forms = get_raw_data_for_workshop(workshop_ids)
+      return get_rollup_for_workshop_ids(
+        workshop_ids,
+        rollup_question_details,
+        false,
+        facilitators
+      )
+    end
+
+    # Given set of facilitators and a course name, return average responses for given
+    # questions across all workshops each facilitator has run.
+    # @param object {facilitator_id: facilitator_name,...} specifying facilitators to include
+    # @param String course_name course name to rollup, ex 'CS Principles'
+    # @param object rollup_question_details questions to include in rollup
+    # @return
+    # {
+    #   general: { see RollupCreator.calculate_averaged_rollup },
+    #   facilitator: { see RollupCreator.calculate_averaged_rollup }
+    # }
+    def self.get_facilitator_rollup_for_course(facilitators, course_name, rollup_question_details)
+      rollups = {general: {}, facilitator: {}}
+      facilitators.each_key do |facilitator_id|
+        workshop_ids = Pd::Workshop.
+          where(course: course_name).
+          where.not(started_at: nil, ended_at: nil).
+          left_outer_joins(:facilitators).where(users: {id: facilitator_id}).distinct.
```

Comment:
Why left join here?

Reply:
good question, when we join with facilitators in the workshop model we always use left outer join--my guess is because it's possible for a workshop to not have a facilitator? But I'm not 100% on that. I will move some of this logic to the workshop model/use workshop model methods.